sumo/src/main.rs

```diff
@@ -0,0 +1,193 @@
+        for lane in &edge.lanes {
+            let lane_id = LaneID(lanes.len());
+            ids_lanes.insert(lane.id.clone(), lane_id);
+            let lane_type = if lane.allow == vec!["pedestrian".to_string()] {
+                LaneType::Sidewalk
+            } else if lane.allow == vec!["bicycle".to_string()] {
+                LaneType::Biking
+            } else if lane.allow == vec!["rail_urban".to_string()] {
+                LaneType::LightRail
+            } else {
+                LaneType::Driving
+            };
+            lanes.push(Lane {
+                id: lane_id,
+                parent: road_id,
+                lane_type,
+                lane_center_pts: lane.center_line.clone(),
+                width: lane.width,
+
+                src_i,
+                dst_i,
+
+                bus_stops: BTreeSet::new(),
+
+                driving_blackhole: false,
+                biking_blackhole: false,
+            });
+            // These seem to appear in the XML from right to left
+            lanes_rtl.push((lane_id, direction, lane_type));
+        }
+
+        if direction == map_model::Direction::Fwd {
+            // Make a new road
+            intersections[src_i.0].roads.insert(road_id);
+            intersections[dst_i.0].roads.insert(road_id);
+            let speed_limit = edge.lanes[0].speed;
+
+            let mut tags = BTreeMap::new();
+            tags.insert("id".to_string(), edge.id.0.clone());
+            if let Some(name) = &edge.name {
+                tags.insert("name".to_string(), name.clone());
+            }
+            let parts: Vec<&str> = edge.edge_type.split(".").collect();
+            // "highway.footway"
+            if parts.len() == 2 {
+                tags.insert(parts[0].to_string(), parts[1].to_string());
+            }
+            let mut lanes_ltr = lanes_rtl;
+            lanes_ltr.reverse();
+
+            roads.push(Road {
+                id: road_id,
+                osm_tags: Tags::new(tags),
+                turn_restrictions: Vec::new(),
+                complicated_turn_restrictions: Vec::new(),
+                orig_id: raw::OriginalRoad::new(123, (456, 789)),
+                speed_limit,
+                access_restrictions: AccessRestrictions::new(),
+                zorder: 0,
+
+                lanes_ltr,
+
+                center_pts: edge.center_line,
+
+                src_i,
+                dst_i,
+            });
+        } else {
+            // TODO Should we check that the attributes are the same for both directions?
+            let mut lanes_ltr = lanes_rtl;
+            lanes_ltr.extend(roads[road_id.0].lanes_ltr.clone());
+            roads[road_id.0].lanes_ltr = lanes_ltr;
+        }
+    }
+
+    let mut internal_lane_geometry: BTreeMap<sumo::LaneID, PolyLine> = BTreeMap::new();
+    for (_, edge) in network.internal_edges {
+        for lane in edge.lanes {
+            if let Some(pl) = lane.center_line {
```

Comment:
Interesting that some lane's dont have a center_line

Reply:
I can't find it now on the wiki, but I read somewhere that `shape` can be elided if it's just the line segment between the two connected nodes. So that's what we infer below.